kyo-core/shared/src/main/scala/kyo/Scope.scala

```diff
@@ -0,0 +1,126 @@
+package kyo
+
+import kyo.Result.Error
+import kyo.Result.Panic
+import kyo.Tag
+import kyo.kernel.ContextEffect
+
+opaque type Scope = Sync
+
+object Scope:
+
+    private val local = Local.init(Maybe.empty[Finalizer])
+
+    def ensure(v: => Any < (Async & Abort[Throwable]))(using Frame): Unit < Scope =
+        ensure(_ => v)
+
+    def ensure(callback: Maybe[Error[Any]] => Any < (Async & Abort[Throwable]))(using Frame): Unit < Scope =
+        withFinalizerUnsafe(f => Abort.get(f.ensure(callback)))
+
+    def acquireRelease[A, S](acquire: => A < S)(release: A => Any < (Async & Abort[Throwable]))(using Frame): A < (Scope & S) =
+        withFinalizerUnsafe { finalizer =>
+            acquire.ensureMap { resource =>
+                val result = finalizer.ensure(_ => release(resource))
```

Comment:
`ensureMap` avoids potential interruptions after `acquire` finishes. Paired with an unsafe `finalizer.ensure`, there's no possibility of interruption between the acquisition finishing and the finalizer callback being registered.